fix(diagnostics): position diagnostics using extmarks

faergeek opened the pull request:
Problem:
Diagnostic positions are not being updated after text changes, which means `vim.diagnostic.open_float` and `vim.diagnostic.jump` will work with outdated positions when text is changed until diagnostics are updated again (if ever).

Solution:
Create extmarks in `vim.diagnostic.set` and use their positions for `vim.diagnostic.open_float` and `next_diagnostic` (used by `vim.diagnostic.jump`, `vim.diagnostic.get_next` and `vim.diagnostic.get_prev`).

### A bit more details

Builtin handlers use extmarks (even if indirectly), so visually it looks fine,
but `open_float` and `jump` use original positions directly causing a mismatch.

Examples:

1. When writing a treesitter query, ftplugin for it has linting on save out of
the box, which sets diagnostics which are then not updated until the next save.
2. LSP diagnostics. Some servers (probably only in very large projects) might
not be able to update diagnostics fast enough.
3. Setting diagnostics from quickfix list. You set diagnostics
`vim.diagnostic.fromqflist` + `vim.diagnostic.set` inside `QuickFixCmdPost`
autocmd, which won't be updated until next time you run `:make`, even though
quickfix list itself usually adjusts positions just fine. It's possible to work
around that by setting diagnostics again in `TextChanged` autocmd, but would be
nicer to turn it into an easier more out-of-the-box experience.

faergeek commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -2150,43 +2277,69 @@ function M.open_float(opts, ...)
 
-  local diagnostics = get_diagnostics(bufnr, opts --[[@as vim.diagnostic.GetOpts]], true)
+  local namespaces = {} ---@type integer[]
+  if type(opts.namespace) == 'number' then
+    namespaces[1] = opts.namespace
+  else
+    for namespace in pairs(diagnostic_cache[bufnr]) do
+      table.insert(namespaces, namespace)
+    end
+  end
```
should `type(opts.namespace) == 'table'` case be supported?

Neither types nor docs seem to allow it, but implicitly it has been supported since
`get_diagnostics` does

MariaSolOs replied:
Opened https://github.com/neovim/neovim/pull/34073 for this.

faergeek replied:
Added it here 3c325a08bdafaf02e7b61fd64634d5f1dcd9842f
Only implementation change.

faergeek commented on test/functional/lua/diagnostic_spec.lua:
```diff
@@ -1104,7 +1104,7 @@ describe('vim.diagnostic', function()
     it('works with diagnostics before the start of the line', function()
       eq(
-        { 4, 0 },
+        { 4, -1 },
```
This has changed because it now returns the original diagnostic that was set without clamping it's position.

Should it be changed? In general if I'm not mistaken it seems the preference is to keep diagnostics unmodified and handle invalid ranges gracefully only when displaying them, based on https://github.com/neovim/neovim/pull/15729

faergeek commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -1179,6 +1242,24 @@ function M.config(opts, namespace)
+---
+---@param bufnr integer Buffer number
+---@param fn fun()
+local function once_buf_loaded(bufnr, fn)
+  if api.nvim_buf_is_loaded(bufnr) then
+    fn()
+  else
+    api.nvim_create_autocmd('BufRead', {
+      buffer = bufnr,
+      once = true,
+      callback = function()
+        fn()
+      end,
+    })
+  end
+end
```
This is useful in particular for `vim.diagnostic.set(vim.diagnostic.fromqflist())` use case.

When using `:make` or similar it's normal to end up with unloaded buffers associated with quickfix items if you didn't edit that buffer before running `:make`, in which case it would not create extmarks and you wouldn't be able to jump between diagnostics.

Builtin diagnostic handlers currently do nothing for unloaded buffers, meaning you won't get underlines, virtual text etc.
Ideally they should also be changed to use this function. I planned to do that in a separate follow-up PR though.

faergeek commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -1007,15 +1009,57 @@ local function next_diagnostic(search_forward, opts)
+      filter_by_severity(opts.severity, diagnostic_cache[bufnr][namespace])
+
+    for _, diagnostic in ipairs(filtered_diagnostics) do
+      for _, extmark in ipairs(extmarks) do
+        if extmark[1] == diagnostic.extmark_id then
+          table.insert(diagnostics, diagnostic)
+        end
+      end
```
This same thing is done in `open_float` as well. Not sure if there should be a separate table mapping extmark ids to diagnostics, which would need to be cleaned up when unloading a buffer or just extracting it into a separate function would be enough.

faergeek commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -1007,15 +1009,63 @@ local function next_diagnostic(search_forward, opts)
-  local line_diagnostics = diagnostic_lines(diagnostics)
+  local line_diagnostics = {} --- @type table<integer,vim.Diagnostic[]>
+  for _, diagnostic in ipairs(diagnostics) do
+    local namespace = diagnostic.namespace
+    assert(namespace, 'namespace must be set')
+
+    local extmark_id = diagnostic.extmark_id
+    assert(extmark_id, 'extmark_id must be set')
```
Seems like these assertions are only needed because `vim.Diagnostic` is used both for an input and output, so that `vim.diagnostic.set`, for example, would not complain about a missing `namespace`.

We should be pretty certain that diagnostics stored in `diagnostic_cache` have both `namespace` and `extmark_id` for sure.

Should I introduce that type separation between input and output?

MariaSolOs replied:
You can combine the assignment and `assert`:
```suggestion
    local namespace = assert(diagnostic.namespace, 'namespace must be set')

    local extmark_id = assert(diagnostic.extmark_id, 'extmark_id must be set')
```

MariaSolOs replied:
Yeah LuaLS isn't the best at complex type inference, but different types would be a bit of an overkill IMO.

faergeek commented:
Not quite sure why luals says that types can't be inferred. I'm not used to luals very much.

For the first one

>runtime/lua/vim/diagnostic.lua:1315:11 [Warning] Can not infer type. (no-unknown)
>              end_col = math.min(end_col, #lines[end_row + 1])
>              ^^^^^^^

Can it be caused by a mismatch between `integer` vs `number` when using `math.min`? Should I add explicit type annotations?

For the other two I have no clue what's wrong at all

>runtime/lua/vim/diagnostic.lua:1318:13 [Warning] Can not infer type. (no-unknown)
>                end_row = end_row - 1
>                ^^^^^^^
>runtime/lua/vim/diagnostic.lua:1323:13 [Warning] Can not infer type. (no-unknown)
>                end_col = end_col - 1
>                ^^^^^^^

Locally with luals in neovim I see the first warning as flaky, sometimes it's there, sometimes it's not. Other two I didn't encounter at all. Maybe something with my setup though, I don't know. Can only see them when running `cmake --build build --target luals`

MariaSolOs commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -1007,15 +1009,63 @@ local function next_diagnostic(search_forward, opts)
-  local line_diagnostics = diagnostic_lines(diagnostics)
+  local line_diagnostics = {} --- @type table<integer,vim.Diagnostic[]>
+  for _, diagnostic in ipairs(diagnostics) do
+    local namespace = diagnostic.namespace
+    assert(namespace, 'namespace must be set')
```
You can combine the assignment and the `assert`:
```suggestion
    local namespace = diagnostic.namespace, 'namespace must be set')
```

faergeek replied:
I guess there's a type in your suggestion?

Did you mean this f8de0867edf33656dc63e4bd5cd5fad6ccccbd00?

MariaSolOs replied:
Yes that. I shouldn't review PRs from my phone lol.

faergeek commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -55,6 +55,8 @@ end
 ---
 --- @field namespace? integer
+---
+--- @field extmark_id? integer
```
Made it private like this 75cb0dc8ab590a2347a1c65b8950410fde3da59d

Had to disable diagnostic in test for luals not to complain. As I understand another way is to prepend `_` to a field name. Which one is preferred?

faergeek replied:
Underscore seems to cause less trouble 754e20563ca019a972c2659955bccaec74374919

MariaSolOs commented:
> @gpanders could you take a look at this PR please?

I would like https://github.com/neovim/neovim/pull/34073 to be merged first though so that this PR doesn't cherry-pick those changes.

faergeek commented:
@gpanders could you take a look? lua_ls doesn't fail anymore, CI is all green.

justinmk commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -2176,44 +2299,81 @@ function M.open_float(opts, ...)
-    end, diagnostics)
+  ---@type vim.Diagnostic[]
+  local diagnostics = {}
+  for _, namespace in ipairs(namespaces) do
```
same here: 1-line comment that gives an intuition of what this nested for-loop is doing.

faergeek replied:
added a comment before the loop and inside

justinmk commented on test/functional/lua/diagnostic_spec.lua:
```diff
@@ -1430,6 +1430,94 @@ describe('vim.diagnostic', function()
       end)
     end)
+
+    it('finds next diagnostic at a logical location after text changes before', function()
```
is it possible for these tests to re-use a shared function, to make them less verbose?

indirection can make things harder to read, but if these tests are mostly doing the same kind of thing, then using a shared function can be much more readable because it makes it obvious what is being tested in each case.

faergeek replied:
reused it via `describe` + `before_each`

justinmk commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -1028,15 +1029,60 @@ local function next_diagnostic(search_forward, opts)
+  local diagnostics = {} ---@type vim.Diagnostic[]
+  ---ns -> extmark_id -> Diagnostic[]
+  local extmarks_index = {} ---@type table<integer, table<integer, vim.api.keyset.get_extmark_item>>
+  for namespace in pairs(diagnostic_cache[bufnr]) do
```
this has triple-nested for-loops and doesn't say even 1 comment giving a hint about what it does.

faergeek replied:
Looking at it now, I think I've messed up and this nesting is completely unnecessary. Currently this loop does two things:

1. Fills in `extmarks_index` with extmarks indexed first by namespace and then extmark id. It's just a cache to avoid calling `nvim_buf_get_extmark_by_id` in multiple loops below (`line_diagnostics` filling, `sort_diagnostics`, `is_next`). I've assumed calling `nvim_buf_get_extmark_by_id` every time instead of looking it up in the cache would negatively affect performance. For no any real reason other than that I saw cache being used on top of `nvim_buf_get_lines` in `get_diagnostics`.
2. Regarding two loops below `filter_by_severity`. I think it should actually just be a concatenation of `filter_by_severity` results across all namespaces and nothing else. I initially wrote it hoping I would be able to get rid of two loops below by using extmark ranges to find next/prev diagnostics, but I couldn't figure out how to properly take severity filtering into account. And I've forgot to clean that up.

At a minimum I will drop these nested loops inside.

I'll also try to figure out again if the loops after this one can be replaced with extmark lookups and if so I think `extmarks_index` will be gone.

But if I won't figure that out, does it make sense to keep `extmarks_index` or should I just call `nvim_buf_get_extmark_by_id` directly where needed instead?

faergeek replied:
I completely removed these loops.

justinmk commented on test/functional/lua/diagnostic_spec.lua:
```diff
@@ -3478,6 +3549,88 @@ describe('vim.diagnostic', function()
       )
     end)
+
+    it('shows diagnostics at their logical locations after text changes before', function()
```
are we actually testing anything here or is this just redundantly testing the existing behavior  of extmarks?

is there a less verbose way to instead (1) assert that extmarks are mapped as expected to a diagnostic, and (2) just assume that extmarks work they way they should?

faergeek replied:
>redundantly testing the existing behavior of extmarks?

In a sense, yes. But just setting extmarks isn't the whole story. If we only check that we set extmarks correctly, we can revert all the changes where they are being used and tests will still happily pass, confirming basically nothing. These tests are more useful to confirm the behavior end to end. They confirm that extmark positions are not only set correctly, but also used to find diagnostics to show them in a floating window or jump between them. For these use cases I don't really see any reason to care that this behavior is implemented using extmarks. That's why extmarks are not mentioned in the tests right now.

These extmarks are not really visible to the user in any way except through API. And testing the behavior the user most likely cares about seems to be more useful than testing an implementation detail, which is only exposed to the user by accident, just because all the extmarks can be easily retrieved by the user through API, no matter what, as far as I'm aware.

But these are just my opinions and I only occasionally contribute and don't maintain the project. So the final decision is not going to be mine anyway.

faergeek commented:
>Does this PR add a new "cache" or is it only updating the existing "cache" structure(s)?

It does add a new local cache for extmarks in `next_diagnostic` (https://github.com/neovim/neovim/pull/34014#discussion_r2144284146). Other than that it adds `_extmark_id` field to an existing `diagnostic_cache`.

>We really don't have code that does something similar already, that can be re-used? Or could make it re-usable somehow? E.g. in `vim.lsp.util` for example.

I'm not very familiar with neovim codebase. But grepping just by `nvim_buf_set_extmark` most usages I've found were in tests. And in the implementation I couldn't find anything similar enough that could've been extracted to be reused here.

faergeek committed:
fix(diagnostics): position diagnostics using extmarks

Problem:
Diagnostic positions are not being updated after text changes, which
means `vim.diagnostic.open_float` and `vim.diagnostic.jump` will work
with outdated positions when text is changed until diagnostics are
updated again (if ever).

Solution:
Create extmarks in `vim.diagnostic.set` and use their positions for
`vim.diagnostic.open_float` and `next_diagnostic` (used by
`vim.diagnostic.jump`, `vim.diagnostic.get_next` and
`vim.diagnostic.get_prev`).

justinmk commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -1247,6 +1292,48 @@ function M.set(namespace, bufnr, diagnostics, opts)
     diagnostic_cache[bufnr][namespace] = diagnostics
   end
 
+  once_buf_loaded(bufnr, function()
```
This is a whole new blob of code that future developers must study. We really need to give hints where we can. In this case, the key idea seems to be (correct if wrong):


```suggestion
  -- Compute positions, set them as extmarks, and store them in
  -- diagnostic._extmark_id (used by get_logical_location). 
  once_buf_loaded(bufnr, function()
```

None of this will be obvious in the future. We need to leave hints.

faergeek commented on runtime/lua/vim/diagnostic.lua:
```diff
@@ -642,6 +643,23 @@ local underline_highlight_map = make_highlight_map('Underline')
+--- @return integer col
+--- @return integer end_lnum
+--- @return integer end_col
+local function get_logical_location(diagnostic)
```
I don't see any problem with that, just wanted to note that I took "logical location" from first paragraph of extmarks docs https://github.com/neovim/neovim/blob/e512efe3696b2ed18db55683fecf42bbc733c767/runtime/doc/api.txt#L476-L481

justinmk replied:
In a description many different terms may be used to orient the reader. But in interface names, same concept should always have same name.
